hw4.js

```diff
@@ -0,0 +1,341 @@
+// (незалежно від того, суворий режим чи ні)
+
+// Якщо значення примітивне (7 або 'foo'), воно конвертується у Object
+// відповідним конструктором - new Number(7) або new String('foo')
+
+// Хм... все одно не зрозуміло, що відбувається
+
+/////////////
+/////////////
+/////////////
+
+// 2
+// https://bigfrontend.dev/quiz/closure-1
+// What does the code snippet to the right output by console.log?
+
+let dev = 'bfe';
+
+function a() {
+  let dev = 'BFE';
+  return function () {
+    console.log(dev);
+  };
+}
+
+dev = 'bigfrontend';
+
+a()(); // 'BFE'
+
+// Таке - a()() - ще не бачив (спойлер - бачив)
+// Якщо говорити про звичайний виклик функції - a() - то гадаю, що
+// виведе 'BFE', бо let dev всередині функції затіняє глобальний let dev,
+// і навіть те, що його значення перезаписане перед викликом, не має сенсу
+
+// Упс, я не помітив вкладену функцію))
+
+// Ну, в принципі те саме повинно бути
+// Чи ні - ми викликаємо функцію, а вона ж поверне нам тоді іншу функцію
+// Здається, я починаю розуміти, чому такий дивний виклик - a()()
+// Я вже вирішував такі завдання на javascript.info, але був невпевненний, що тут та сама ситуація
+// Виходить, ми викликаємо зовнішню функцію, а потім внутрішню
+// Тоді внутрішня функція візьме значення dev з ОВ зовнішньої і поверне 'BFE'
+// Тобто результат той самий, але механізм інший
+
+/////////////
+/////////////
+/////////////
+
+// 3
+// https://bigfrontend.dev/quiz/comparison
+// What does the code snippet to the right output by console.log?
+
+console.log(10 > 9); // true
+console.log(10 > '9'); // true, бо відбудеться приведення до типу
+console.log('10' > 9); // true, бо відбудеться приведення до типу
+console.log('10' > '9'); // false, бо це різні строки
+
+// Про всяк випадок додам, що в одному з випадків приведення до типу
+// може не відбутись (або не так відбутись) і тоді буде false
+// Але чому і який це буде випадок важко сказати
+// Якщо припустити, що змінює тип завжди друге значення,
+// тоді у передостанньому випадку теж буде false,
+// бо у такому разі це фактично той самий вираз, що й останній
+
+// Дарма вагався, вгадав з першого разу))
+
+/////////////
+/////////////
+/////////////
+
+// PRACTICE
+
+// 1
+// Implement a function setCancellableInterval, that acts like setInterval but
+// instead of returning a timer ID, it returns a function that when called,
+// cancels the interval. The setCancellableInterval function should have the
+// exact same signature as setInterval:
+
+// Example:
+
+let i = 0;
+// t = 0:
+
+const cancel = setCancellableInterval(() => {
+  i++;
+}, 10);
+// t = 10: i is 1
+// t = 20: i is 2
+
+cancel(); // Called at t = 25
+
+// t = 30: i is still 2 because cancel() was called and
+// the interval callback has stopped running.
+
+//Якщо я правильно зрозумів умову, тоді так:
+
+function setCancellableInterval() {
+  return function clearInt() {
+    clearInterval();
+  };
+}
+
+// Але щось мені підказує, що я зрозумів неправильно
```

Comment:
подивись ще раз на приклад

```
const cancel = setCancellableInterval(() => {
  i++;
}, 10);

cancel();
```

в тебе є функція setCancellableInterval яка запускає під капотом [setInterval](https://developer.mozilla.org/en-US/docs/Web/API/setInterval), але окрім цього ще й повертає функцію, яку можна викликати, й цей інтервал перестане запускатись.

Іншими словами, за допомогою звичайного setInterval твій код виглядав би так:
```
const someIntervalId = setInterval(() => {
  console.log('цей код буде запускатись поки не відмінеться інтервал')
}, 100)

// відміняю інтервал
clearInterval(someIntervalId)

```

а за допомогою твоєї функції він буде виглядати ось так:

```

const cancel = setCancellableInterval(() => {
  console.log('цей код буде запускатись поки не відмінеться інтервал')
}, 100);

// очищаю інтервал
cancel();
```

Твоє рішення містить гарний код, але наразі в тебе clearInterval викликається без передачі intervalId й він нічого не буде очищати

Reply:
Над цим ще треба посидіти, поки не доходить...

Reply:
```

let i = 0;
// t = 0:
const cancel = setCancellableInterval(() => {
  i++;
}, 10);
// t = 10: i is 1
// t = 20: i is 2
cancel(); // Called at t = 25
// t = 30: i is still 2 because cancel() was called and the interval callback has stopped running.
```